File: LeetCode Mastery/1. Two Sum/two-sum/src/Solution_Three.java
import java.util.Arrays;
import java.util.HashMap;

/*
 * Solution Three
 * - single pass HashMap
 * - O(n)
 * - runtime: 2 ms
 * - memory: 44.79 MB
 */
public class Solution_Three {
    public Solution_Three(int[] input_nums, int input_target) {
        System.out.println("\nSolution 3 -->");
        System.out.println("nums[]: " + Arrays.toString(input_nums));
        System.out.println("target: " + input_target);
        System.out.println("output: " + Arrays.toString(twoSum(input_nums, input_target)));
    }
    
    private int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> nums_map = new HashMap<>();
        nums_map.put(nums[0], 0);
        for (int i = 1; i < nums.length; i++) {
            int complement = target - nums[i];
            if (nums_map.containsKey(complement) && (nums_map.get(complement) != i)) {
                return new int[] {i, nums_map.get(complement)};
            } else {
                nums_map.put(nums[i], i);
            }
        }
        return null;
    }
}

File: LeetCode Mastery/1. Two Sum/two-sum/src/Solution_Three.py
# Solution 3
# - single pass HashMap
# - O(n)
# - runtime: 48 ms
# - memory: 17.80 MB

from typing import List

class Solution_Three:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        n = len(nums)
        nums_map = {}
        nums_map[nums[0]] = 0
        for i in range(1, n):
            complement = target - nums[i]
            if complement in nums_map and nums_map[complement] != i:
                return [i, nums_map[complement]]
            else:
                nums_map[nums[i]] = i

    nums = [2, 7, 11, 15]
    target = 9
    print(twoSum(self = True, nums = nums, target = target))